feat(Airtable Node): Add Option to Use Field ID's

pnogemane opened the pull request:
Feature Specification - The request was an enhancement for the [Airtable node](https://docs.n8n.io/integrations/builtin/app-nodes/n8n-nodes-base.airtable/) in n8n. Currently, the node requires field names when creating or updating records, but Airtable’s API uses field IDs. This can cause issues, especially when field names change or contain special characters, breaking workflows.

A similar functionality for targeting tables by ID instead of table name has been discussed and implemented (see [this thread](https://community.n8n.io/t/airtable-node-target-table-with-table-name-instead-of-id/26067)), and having a similar feature for fields would be extremely useful. There was a similar feature request [here 1](https://community.n8n.io/t/airtable-node-option-to-allow-for-using-field-ids/29298), but it hasn’t received much traction.

Proposed Solution:

Add an option for the Airtable node to accept field IDs instead of field names for creating and updating records.
In the n8n UI, the fields should still display the field names for easy configuration, but the underlying logic should use the field IDs, making workflows more resilient to Airtable schema changes.

## Summary

<!--
Describe what the PR does and how to test.
Photos and videos are recommended.
-->


## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [x] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [x] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

pnogemane committed:
Add Option to Use Field ID's in Airtable Node

Updated getColumns method to retrieve field IDs while maintaining backward compatibility.
This method is now reusable for both create and update functions.
Introduced a new getFieldID method in loadOptions to satisfy task requirements.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=11666) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=11666) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=11666) it.</sub>

pnogemane commented:
close #bluedotimpact/bluedot#19

pnogemane commented:
close #bluedotimpact/bluedot#19

pnogemane commented:
close #bluedotimpact/bluedot#19

tarinrickett commented on packages/nodes-base/nodes/Airtable/v2/methods/loadOptions.ts:
```diff
@@ -25,13 +25,22 @@ export async function getColumns(this: ILoadOptionsFunctions): Promise<INodeProp
-			name: field.name as string,
-			value: field.name as string,
-			description: `Type: ${field.type}`,
+			name: fieldLabel as string,
```
why are we replacing name with id if `useFieldIDs` is true? I am concerned the naming will become confusing. and won't we always need the `name` for the UI, even if `useFieldIDs` is true?

```suggestion
			name: field.name as string,
			id: field.id as string,
```

pnogemane replied:
Resolved in commit 

commit 6093ab98d1ecccf3d901625c0a4b5bfa6733c19c 

Description 

Updated the retrieval of baseId and tableId by using undefined as the fallback value, ensuring clarity in cases where these parameters are missing. If either baseId or tableId is not provided, a NodeOperationError is thrown to prevent further execution with incomplete data. This enhances error handling by explicitly requiring both parameters for correct function operation.

tarinrickett reviewed (commented):
first time reviewing in this codebase! a few clarifying questions + suggestions.

pnogemane committed:
Line 29 : Managing Defaults

A DEFAULTS object has been introduced to centralize default values, including the default for useFieldIDs. The value of useFieldIDs is now retrieved using this.getNodeParameter(), and if not provided, it defaults to the value defined in DEFAULTS.useFieldIDs. This change enhances code maintainability by consolidating default values in a single location, making it easier to update and manage them in the future. Additionally, this approach ensures backward compatibility and avoids assuming that the third parameter will default to undefined when not passed in, allowing for better control over default behavior.

pnogemane committed:
Updated the retrieval of baseId and tableId by using undefined as the fallback value

Updated the retrieval of baseId and tableId by using undefined as the fallback value, ensuring clarity in cases where these parameters are missing. If either baseId or tableId is not provided, a NodeOperationError is thrown to prevent further execution with incomplete data. This enhances error handling by explicitly requiring both parameters for correct function operation.

pnogemane committed:
Handling Field ID's separately

The code now includes a conditional check for useFieldIDs. If useFieldIDs is true, we assign field.id to fieldId, otherwise, fieldId is set to undefined. This keeps the label (name) the same for UI purposes while optionally using fieldId for backend operations. The result is populated with these values, ensuring compatibility and flexibility in the code without changing the overall structure

pnogemane committed:
Get Field ID's Documentation

/**
 * Fetches the field IDs for a specific table in a given base.
 *
 * This function retrieves field information from the Airtable API and returns a mapping of field names to their corresponding field IDs.
 * It is primarily used for mapping fields in Airtable tables, allowing users to reference fields by their IDs rather than names.
 *
 * @param {ILoadOptionsFunctions} this - The context in which the function is called, providing access to node parameters.
 * @returns {Promise<Record<string, string>>} A promise that resolves to a record containing field names as keys and their corresponding field IDs as values.
 * @throws {NodeOperationError} Throws an error if the response from the Airtable API is invalid or does not contain the expected data structure.
 *
 * @example
 * const fieldIds = await getFieldIds.call(this);
 * console.log(fieldIds); // { "Name": "fld12345", "Email": "fld67890", ... }
 */

pnogemane commented on packages/nodes-base/nodes/Airtable/v2/methods/loadOptions.ts:
```diff
@@ -25,13 +25,22 @@ export async function getColumns(this: ILoadOptionsFunctions): Promise<INodeProp
 	}
 
+
+	const useFieldIDs = this.getNodeParameter('useFieldIDs', undefined, false) as boolean;
```
Resolved in 

commit 58caf4185cc3464f5a97b4a399f55454e29c8878

Commit Link : https://github.com/n8n-io/n8n/pull/11666/commits/58caf4185cc3464f5a97b4a399f55454e29c8878 

Description 

A DEFAULTS object has been introduced to centralize default values, including the default for useFieldIDs. The value of useFieldIDs is now retrieved using this.getNodeParameter(), and if not provided, it defaults to the value defined in DEFAULTS.useFieldIDs. This change enhances code maintainability by consolidating default values in a single location, making it easier to update and manage them in the future. Additionally, this approach ensures backward compatibility and avoids assuming that the third parameter will default to undefined when not passed in, allowing for better control over default behavior.

pnogemane commented on packages/nodes-base/nodes/Airtable/v2/methods/loadOptions.ts:
```diff
@@ -101,3 +110,52 @@ export async function getAttachmentColumns(
+
+export async function getFieldIds(this: ILoadOptionsFunctions): Promise<Record<string, string>> {
+	const baseId = this.getNodeParameter('baseId', 0) as string;
+	const tableId = encodeURI(this.getNodeParameter('tableId', 0) as string);
```
Resolved in commit 

commit 938b91c4c1383d58afae62c5dfec8a28b14b6cae

Description 

Updated the retrieval of baseId and tableId by using undefined as the fallback value, ensuring clarity in cases where these parameters are missing. If either baseId or tableId is not provided, a NodeOperationError is thrown to prevent further execution with incomplete data. This enhances error handling by explicitly requiring both parameters for correct function operation.

pnogemane commented on packages/nodes-base/nodes/Airtable/v2/methods/loadOptions.ts:
```diff
@@ -101,3 +110,52 @@ export async function getAttachmentColumns(
+	name: string;
+}
+
+export async function getFieldIds(this: ILoadOptionsFunctions): Promise<Record<string, string>> {
```
Resolved In Commit 

commit a115d229a4d5084bf60a3ad6a39fbf7956d0aa3b

Get Field ID's Documentation
/**
 * Fetches the field IDs for a specific table in a given base.
 *
 * This function retrieves field information from the Airtable API and returns a mapping of field names to their corresponding field IDs.
 * It is primarily used for mapping fields in Airtable tables, allowing users to reference fields by their IDs rather than names.
 *
 * @param {ILoadOptionsFunctions} this - The context in which the function is called, providing access to node parameters.
 * @returns {Promise<Record<string, string>>} A promise that resolves to a record containing field names as keys and their corresponding field IDs as values.
 * @throws {NodeOperationError} Throws an error if the response from the Airtable API is invalid or does not contain the expected data structure.
 *
 * @example
 * const fieldIds = await getFieldIds.call(this);
 * console.log(fieldIds); // { "Name": "fld12345", "Email": "fld67890", ... }
 */

pnogemane commented on packages/nodes-base/nodes/Airtable/v2/methods/loadOptions.ts:
```diff
@@ -101,3 +110,52 @@ export async function getAttachmentColumns(
+					}
+
+					if (map[field.name]) {
+
```
The formatting issue has been updated and committed as part of the list of commits, in accordance with the review requirements.

netroy commented on packages/@n8n/permissions/package.json:
```diff
@@ -19,5 +19,8 @@
     "dist/**/*"
-  ]
+  ],
+  "dependencies": {
```
what was the reason for this change?

pnogemane replied:
The change was initially automated to track the introduction of a new file. However, the file has since been removed.

pnogemane reviewed (commented):
Feature: Add Option to Use Field IDs in Airtable Node ---- Pull Request Completion 

Summary: These updates where required as part of the pull request to complete implementation of the "Add Option to Use Field IDs in Airtable Node" feature.

pnogemane committed:
Merge branch 'n8n-io:master' into master

pnogemane committed:
Merge pull request #1 from n8n-io/master

Merging Branch To Master

pnogemane committed:
Merge branch 'n8n-io:master' into master

pnogemane committed:
Merge pull request #2 from n8n-io/master-n8n-Add-option-to-use-field-IDs-in-Airtable-node

n8n-Add-option-to-use-field-IDs-in-Airtable-node

/close #19

pnogemane commented:
Merged

 See pull Request : https://github.com/pnogemane/n8n-Add-option-to-use-field-IDs-in-Airtable-node/pull/2

tarinrickett commented:
hi @netroy 👋 I saw you gave this an earlier first pass – friendly bump; can you take another look?

our team set this bounty and is interested in getting this fix merged, but are not n8n owners to approve the PR ourselves. let us know if there's anything we can do to support! thanks!!

netroy commented on packages/@n8n/permissions/package.json:
```diff
@@ -19,5 +19,8 @@
     "dist/**/*"
-  ]
+  ],
+  "dependencies": {
```
This change is still here. Please revert this.

netroy commented on packages/nodes-base/nodes/Airtable/Airtable.node.ts:
```diff
@@ -4,6 +4,7 @@ import { VersionedNodeType } from 'n8n-workflow';
 import { AirtableV1 } from './v1/AirtableV1.node';
 import { AirtableV2 } from './v2/AirtableV2.node';
 
+
```
not sure what this additional empty line is for.

netroy commented:
> hi @netroy 👋 I saw you gave this an earlier first pass – friendly bump; can you take another look?

I'm not the code-owner for this part of the codebase. The nodes team will pick this up as soon as they can, as they have a lot context in this case.

tarinrickett commented:
> > hi @netroy 👋 I saw you gave this an earlier first pass – friendly bump; can you take another look?
> 
> I'm not the code-owner for this part of the codebase. The nodes team will pick this up as soon as they can, as they have a lot context in this case.

ah got it, thanks!

tarinrickett commented:
hi @pnogemane , are you able to complete the unresolved comments and facilitate an owner's approval so we can get this merged?

we won't pay [the bounty](https://github.com/bluedotimpact/bluedot/issues/19) until this fix is merged with the main branch and confirmed to work, and the bounty will go to whatever developer is able to complete the task first.

hope you are able to follow up so we can close this up quickly. let us know if you have any outstanding questions.

Joffcom commented:
Hey @pnogemane,

Can you resolve the requested comments, Once you have done this I will be able to give this another quick test to make sure it doesn't break any existing workflows and we can get it merged.
